### PR TITLE
ref: corrections to `exp run/apply/branch` +

### LINKED ISSUE #2399: Improve explanation of error handling for interrupting checkpoints
See https://github.com/iterative/dvc-checkpoints-mnist/pull/2 for background.

Here's the scenario:

1. My script includes a checkpoint (using `make_checkpoint()` or similar) but no error handling for keyboard interuptions.
2. I run an experiment and interrupt it with ctrl-c, which returns a non-0 exit code.
3. I resume the experiment.

If there were changes between the last time `make_checkpoint()` was called and when the experiment was interrupted, step 3 will create a new experiment because there was no final cleanup checkpoint applied at the end of step 2 (if the script exited cleanly with a 0 exit code, there would be a final cleanup checkpoint).

I couldn't figure this out until @pmrowla explained it, so IMO it's not going to be clear to users. Since we recommend this ctrl-c workflow in the docs and in our examples, I think we need to better document it. Thoughts?

### PR DESCRIPTION
**Pull request by jorgeorpinel**
- Addresses https://github.com/iterative/dvc.org/pull/2243#pullrequestreview-603422799 and https://github.com/iterative/dvc.org/pull/2286#pullrequestreview-615217332 (finally)
- And 1. from #2397 and close #2399 (inc. `make_checkpoint()` update)
- Other improvements and copy edits

### CODE REVIEW

File: content/docs/command-reference/exp/branch.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/branch.md:23]: I don't think there's anything wrong with this sentence, but I found it a little confusing. Is this necessary? Seems more like a general explanation of experiments than anything specific to `exp branch`, but I might be missing the point you're trying to convey.

File: content/docs/command-reference/exp/branch.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/branch.md:27]: I would say there are a few other uses:
1. Wanting to preserve multiple experiments by creating a branch from each.
2. Wanting to preserve experiments without needing to apply them to the workspace (if you're in the middle of experimenting with other things).
3. Wanting to preserve checkpoint commits in your Git history (instead of squashing them like `exp apply` would do).
It's not necessary to be exhaustive, but I actually think that `exp branch` primarily for these reasons rather than as a backup to `exp apply`.

File: content/docs/command-reference/exp/branch.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/branch.md:30]: My vote is still to drop this since it's slightly inaccurate, and I don't think it adds much that's not already covered by the description, but I know we discussed it in a previous PR, so 🤷 .

File: content/docs/command-reference/exp/branch.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/branch.md:40]: 🔥

File: content/docs/command-reference/exp/apply.md
Comment (pmrowla) [reply content/docs/command-reference/exp/apply.md:27]: It fails outright if your current `HEAD` commit does not match the `HEAD` commit from when you ran the experiment.

It may also fail if there are git conflicts between changes you have in your workspace and changes in the experiment you want to apply. We try to automatically resolve the conflicts by preferring the changes in the applied experiment, but sometimes git may still fail to auto-resolve them.

Comment (jorgeorpinel) [reply content/docs/command-reference/exp/apply.md]: Updated in 8762f974. PTAL

> We try to automatically resolve the conflicts by preferring the changes in the applied experiment

Meaning `git merge --theirs`, right @pmrowla ? I think it's not considered a conflict until after Git can't merge it so I went with the phrase "if they conflict", for now.

> Where  they have moved their HEAD ... and then try to exp apply some experiment

Does/should the CLI print a HINT to try `exp branch` instead?

File: content/docs/command-reference/exp/branch.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md:23]: Yeah maybe it doesn't belong in the first paragraph. But I think it's key that people understand that the branch won't be based on their current `HEAD`. BTW the sample `git` commands below are still incorrect, I see... ⌛

Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md]: Changed a bit 2685e741 for now, but I may change it per your other feedback...

File: content/docs/command-reference/exp/apply.md
Comment (pmrowla) [reply content/docs/command-reference/exp/apply.md]: No, it will give an error along the lines of "cannot apply `exp-...` onto commit `<user's current HEAD>`. Experiment is derived from `<original experiment HEAD>`"

File: content/docs/command-reference/exp/branch.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md]: Yeah, very good summary! I put this info in (with some addendums) in 2510027e, but still keeping the `exp apply` contrasting 😬

File: content/docs/command-reference/exp/branch.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md]: Hmmm yeah. I think you're right. ✖️'d

File: content/docs/command-reference/exp/branch.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md]: 👍

File: content/docs/command-reference/exp/apply.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/apply.md]: > exp branch is for a different use case than exp apply

OK @pmrowla. Would you mind checking the bullet list in the latest version to confirm that it covers the use case(s)? Thanks

File: content/docs/command-reference/exp/branch.md
Comment (pmrowla) [inline content/docs/command-reference/exp/branch.md:35]: This case doesn't make sense to me, `exp branch` is not a replacement for `exp apply`. `exp branch` is for creating a Git branch that contains all of the commits in an experiment run.

`exp apply` is for applying the (merged) contents of a single experiment commit into my current workspace.

File: content/docs/command-reference/exp/apply.md
Comment (iesahin) [inline content/docs/command-reference/exp/apply.md:17]: Could we make the statement in parentheses a new sentence?

Restores a given `experiment` into the workspace. Experiment can be referenced by a hash or name.

File: content/docs/command-reference/exp/branch.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/branch.md:27]:
```suggestion
time that the experiment was run).
```
`git branch` also doesn't switch to the new branch, right? Is the point of this paragraph to clarify that the current HEAD/branch has no relation to `exp branch`?

File: content/docs/command-reference/exp/run.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/run.md:91]: It shouldn't start from scratch, it will just potentially start a modified experiment if there are "dirty" changes left over that happened after the last checkpoint was created.

File: content/docs/command-reference/exp/branch.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md:35]: How would you "apply" a past non-checkpoint experiment (once `HEAD` has moved)?

File: content/docs/command-reference/exp/apply.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/apply.md:17]: Doesn't seem like a sentence pertaining to that paragraph to me. Would have to be another p just for the argument (and maybe some of the options), which I'm trying to avoid.

File: content/docs/command-reference/exp/branch.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md:27]: Ah yeah I was thinking `git checkout -b` I guess. Fixed

The paragraph clarifies where the branch is created, yes. Since the fact that exps are already commits is hidden from the user, it shouldn't be obvious to them that this just relabels an existing custom Git ref. or what it's tree is.

File: content/docs/command-reference/exp/run.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/run.md]: OK, I wasn't sure about this... (removed that last sentence for now). In #2399 you mentioned:

> step 3 will create a new experiment because there was no final cleanup checkpoint

But I didn't test what this looks exactly. What's a "new experiment" here, if it does continue from the last checkpoint?

File: content/docs/command-reference/exp/branch.md
Comment (pmrowla) [reply content/docs/command-reference/exp/branch.md:35]: You can't. The only way you can use `dvc exp apply` in this scenario is with
```
git checkout <old-head-commit>
exp apply <exp>
```

If by "apply" here you really mean that you want to merge or rebase the changes from an experiment onto your new HEAD, you have to do it manually with the appropriate git commands. see also: https://github.com/iterative/dvc/issues/5782

Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md:35]: What about something like
```console
$ dvc exp branch exp-12341 exp-branch
$ git merge exp-branch
```
? (half DVC half Git)

Comment (pmrowla) [reply content/docs/command-reference/exp/branch.md:35]: yes that will work, but it has the side effect of creating a full Git branch containing your experiment. If you have a ton of experiments you want to test against your new HEAD in this way, this is not ideal, since you probably don't actually want full Git branches for every one of those experiments.

Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md:35]: For now maybe
```suggestion
- In cases where it's not possible to `dvc exp apply` an experiment directly;
```

Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md]: Rephrased in 68892454. WDYT? If it feels to forced, obvious, or indirect we could just remove it indeed. But then there's only 2 items in this bullet list and they kind of relate so would need to rewrite as a single scenario/motivation paragraph.

Comment (pmrowla) [reply content/docs/command-reference/exp/branch.md]: I still don't really think that bullet point belongs in the list.

Honestly it would probably be better to just combine the last two bullet points into a single statement/paragraph as you suggested.

It should just be something like

> `dvc exp branch` is useful if you would like to an experiment persistent as a Git branch, and for preserving checkpoints as independent commits within that newly created Git branch.

File: content/docs/command-reference/exp/run.md
Comment (dberenbaum) [reply content/docs/command-reference/exp/run.md]: Take a look at https://github.com/iterative/dvc-checkpoints-mnist/pull/2. That should help illustrate what happens, which is the same thing that would happen if someone changed parameters or code before running `exp run` again. 

`(it will restart from scratch)` will be misleading here I think.

File: content/docs/command-reference/exp/apply.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/apply.md:26]:
```suggestion
Note that any intermediate [checkpoints](/doc/command-reference/exp/run#checkpoints) found in
```
Wondering if `intermediate` better conveys the message that the whole history is "squashed" into a single merge rather than disregarded. `apply` will still reflect the experiment state at the time of the final checkpoint. Feel free to ignore if you think it's unhelpful.

File: content/docs/command-reference/exp/branch.md
Comment (dberenbaum) [reply content/docs/command-reference/exp/branch.md]: IMO the first bullet can be put into the above paragraph and the last two bullets removed. I think the `apply` comparison is a bit too much of a stretch, and the independent checkpoint commits are mentioned already above. There's no reason the use cases need to be exhaustive. Giving the primary usage seems natural to me.

Comment (jorgeorpinel) [reply content/docs/command-reference/exp/branch.md]: OK I made it into a regular p with only the first item (rephrased slightly).

File: content/docs/command-reference/exp/apply.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/apply.md:17]: I agree it would be easier to read though...

File: content/docs/api-reference/make_checkpoint.md
Comment (dberenbaum) [inline content/docs/api-reference/make_checkpoint.md:86]:
```suggestion
        exit()
```

File: content/docs/command-reference/exp/apply.md
Comment (dberenbaum) [inline content/docs/command-reference/exp/apply.md:27]: Alternate suggestion (same idea as previous comment adding `intermediate`):
```suggestion
the `experiment` are **not** preserved in separate commits when applying.
```
Again, not a blocker, decide for yourself.

**Review by dberenbaum — APPROVED** [review]
Left very minor suggestions to consider, but looks ready to merge.

File: content/docs/command-reference/exp/run.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/run.md]: Hm OK I see. I had to intro the concept of "wrap-up checkpoint" to explain this. See 1ceeebef

File: content/docs/command-reference/exp/apply.md
Comment (jorgeorpinel) [reply content/docs/command-reference/exp/apply.md:26]: Yeah it's more correct, but not really a concept we have explained much. I don't know if it's intuitive enough. I went for "the history of checkpoints" for now.

**Pull request merged**
